packages/@lwc/babel-plugin-component/src/decorators/index.js

```diff
@@ -163,39 +139,133 @@ function generateInvalidDecoratorError(path) {
     }
 }
 
+function collectDecoratorPaths(bodyItems) {
+    return bodyItems.reduce((acc, bodyItem) => {
+        const decorators = bodyItem.get('decorators');
+        if (decorators && decorators.length) {
+            acc.push(...decorators);
+        }
+        return acc;
+    }, []);
+}
+
+function getDecoratorMetadata(decoratorPath) {
+    const expressionPath = decoratorPath.get('expression');
+
+    let name;
+    if (expressionPath.isIdentifier()) {
+        name = expressionPath.node.name;
+    } else if (expressionPath.isCallExpression()) {
+        name = expressionPath.node.callee.name;
+    } else {
+        throw generateInvalidDecoratorError(decoratorPath);
+    }
+
+    const propertyName = decoratorPath.parent.key.name;
+    const binding = decoratorPath.scope.getBinding(name);
+    const type = getDecoratorType(decoratorPath.parentPath);
+
+    return {
+        name,
+        propertyName,
+        binding,
+        path: decoratorPath,
+        type,
```

Comment:
The term `type` is confusing, I would have expected to refer to the type of decorator while it refers to the type of class property. What about renaming this property to `classPropertyType` and `getDecoratorType` into `getClassPropertyType`?

Reply:
I went with `type` because I wanted to steer away from variable names like `classPropertyOrMethodType`...what do you think about `decorateeType` or `decoratedType`?

Reply:
`decoratedType` type is a good idea. I was also thinking about `decoratedNodeType`.